### PR TITLE
feat(cli): Regularize enarx cpu info output

### LINKED ISSUE #1962: Regularize `enarx cpu info` output
Currently the default `enarx cpu info` output looks like this:

```
✗ Backend: sgx
  ✗ Driver: /dev/sgx_enclave
  ✗ AESM Daemon Socket: /var/run/aesmd/aesm.socket
  ✔ CPU: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz | GenuineIntel
  ✗  SGX Support
  ✗   Version 1
  ✗   Version 2
  ✗   FLC Support
  ✔   Max Size (32-bit): 1
  ✔   Max Size (64-bit): 1
  ✔   MiscSelect: (empty)
  ✔   Features: (empty)
  ✔   Xfrm: (empty)
  ✔   EPC Size: 0
```

Which isn't totally ideal, but is fine enough since this is just meant for human consumption.

However, if a user asks for JSON, they get stuff this:

```
      "backend": "sgx",
      "data": [
        {
          "name": "Driver",
          "pass": false,
          "info": "/dev/sgx_enclave",
          "mesg": null
        },
        {
          "name": "AESM Daemon Socket",
          "pass": false,
          "info": "/var/run/aesmd/aesm.socket",
          "mesg": null
        },
        {
          "name": "CPU",
          "pass": true,
          "info": "11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz | GenuineIntel",
          "mesg": null
        },
        {
          "name": " SGX Support",
          "pass": false,
          "info": null,
          "mesg": null
        },
        {
          "name": "  Version 1",
          "pass": false,
          "info": null,
          "mesg": null
        },
```

Some obvious problems: whereas the output pretends to be a hierarchy, the structure of the hierarchy is represented here by indenting the value of the name field. Furthermore, whereas consumers of this data would probably prefer to have proper key/value pairs, for most of this data all the values are just stored directly in the name field, with the info field being null.

Figure out a better representation for the JSON output to make it more regular to process and consume.

### PR DESCRIPTION
**Pull request by gandhisamay**
Formatted the CPU info output for the json option.

Fixes: https://github.com/enarx/enarx/issues/1962

### CODE REVIEW

**Review by haraldh — CHANGES_REQUESTED** [review]

File: src/backend/mod.rs [hunk 1]
```diff
@@ -184,14 +186,78 @@ pub trait Backend: Sync + Send {
...
+        }
+
+        fn rec_insert(mut val: &mut Value, datum_json: Value, counter: i32) {
+            //each unwrap will never fail
+            if counter != 1 {
+                rec_insert(val.get_mut("data").unwrap(), datum_json, counter - 1)
+            } else {
+                let is_object = val.is_object();
+
+                if is_object {
```
Comment (haraldh) [inline src/backend/mod.rs:208]:
```suggestion
                if val.is_object() {
```

File: src/backend/mod.rs [hunk 2]
```diff
@@ -184,14 +186,78 @@ pub trait Backend: Sync + Send {
...
+                        last_prop.push(datum_json);
+                    } else {
+                        val = last_prop.last_mut().unwrap();
+                        rec_insert(val, datum_json, counter)
+                    }
+                }
+            }
+        }
+
+        //Logic
```
Comment (haraldh) [inline src/backend/mod.rs:229]: Logic what?

**Review by haraldh — DISMISSED** [review]

**Comment by haraldh** [comment]
Don't use merge commits and squash your commits into logical ones. Don't have commits only for formatting.

**Comment by haraldh** [comment]
Thanks for submitting this PR!

File: src/backend/mod.rs (hunk 1 above)
Comment (gandhisamay) [reply src/backend/mod.rs:208]: Noted!

File: src/backend/mod.rs (hunk 2 above)
Comment (gandhisamay) [reply src/backend/mod.rs:229]: Just for anyone trying to understand the code, that the overall logic can be seen from here.

**Comment by gandhisamay** [comment]
I have made all the suggested changes.

**Comment by haraldh** [comment]
Thanks! Please squash all your commits into one. That is sufficient for this PR.

**Comment by gandhisamay** [comment]
> Thanks! Please squash all your commits into one. That is sufficient for this PR.

Done!!

**Comment by nickvidal** [comment]
Thank you @gandhisamay!

**Comment by bstrie** [comment]
This looks good to me. I do have two comments:

1. The approach here determines the depth of the JSON nesting based on counting leading whitespaces. However, those leading whitespaces were already determined elsewhere based on an existing hierarchy of data. In the long term, I would prefer to *only* insert whitespaces in the "pretty" output, and the JSON should be generated directly from the underlying data instead. However, that's not a user-visible change, so I'm fine leaving that for the future.
2. I'd like to have a test for the output of `enarx platform info`. The problem is that the output of this command is highly platform-specific. However, we just need the test to ensure the general structure of the JSON output, and we don't actually care about the platform info itself. This means that we could generate a dummy backend as a part of the test, implement `Backend` for it, and then have a test that serializes the dummy backend and checks the output. I'm fine leaving this for the future, but if @gandhisamay is looking for something to work on next, this would be a good place to start. :)

**Review by bstrie — APPROVED** [review]

**Comment by bstrie** [comment]
Ah, we might need to update the nix flake in order to get the CI to pass.

**Comment by gandhisamay** [comment]
@bstrie 
1)  The main problem is that the struct `Datum` in itself doesn't represent the actual hierarchy and the whitespaces are actually the ones which represent that hierarchy currently for a particular backend, so maybe what we could do is just like the `data` field present in the json structure if we have a vector of datum present in the datum struct itself then we would spare ourselves from the all this whitespaces related logic and the json output serialization will as well simply significantly and seems correct for a longer as well. (If agreed upon by everyone, I'm willing to implement this)
2) Do we want the test with a dummy backend as a part of this PR itself or create a separate issue for that?

**Comment by bstrie** [comment]
> If agreed upon by everyone, I'm willing to implement this

If you'd be willing to try, then go ahead. :)

> Do we want the test with a dummy backend as a part of this PR itself or create a separate issue for that?

If you'd like to do this right now, then feel free to make it a part of this PR. However, if you're busy with other things, then we can leave that for the future and merge this PR. I would make this a higher priority than the Datum refactor.

**Comment by gandhisamay** [comment]
> @gandhisamay
> 
> > If agreed upon by everyone, I'm willing to implement this
> 
> If you'd be willing to try, then go ahead. :)
> 
Cool, I will give it a try
> > Do we want the test with a dummy backend as a part of this PR itself or create a separate issue for that?
> 
> If you'd like to do this right now, then feel free to make it a part of this PR. However, if you're busy with other things, then we can leave that for the future and merge this PR. I would make the test a higher priority than the Datum refactor.
Okay then, I will write the test now and update this PR

**Comment by gandhisamay** [comment]
@bstrie I have implemented the dummy backend test, could you please review it?

**Review by bstrie — APPROVED** [review]

**Pull request merged**